### PR TITLE
Implementing s3

### PR DESCRIPTION
**Pull request by joeylouise**
This PR does 3 things

- When user adds memory, pushes to S3
- When view memory, pulls from S3
- When deleting, deletes from S3

#116 #125

### CODE REVIEW

File: public/scripts/s3-uploads.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
+(function () {
+  document.getElementById('camera-input').onchange = function () {
+    const files = document.getElementById('camera-input').files;
+    const file = files[0];
+    if (file == null) {
+      return alert('No file selected.');
```
Comment (des-des) [inline public/scripts/s3-uploads.js]: ? Is this good UX?

File: public/scripts/s3-uploads.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    getSignedRequest(file);
+  };
+}());
+
+function getSignedRequest(file) {
+  const xhr = new XMLHttpRequest();
+  xhr.open('GET', `/sign-s3?file-name=${file.name}&file-type=${file.type}`);
+  xhr.onreadystatechange = () => {
```
Comment (des-des) [inline public/scripts/s3-uploads.js]: Create a function in helpers called request. Use it here

File: public/scripts/s3-uploads.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    if (xhr.readyState === 4) {
+      if (xhr.status === 200) {
+        const response = JSON.parse(xhr.responseText);
+        uploadFile(file, response.signedRequest, response.url);
+        saveInLocalStorage(response.key);
+      } else {
+        alert('Could not get signed URL.');
```
Comment (des-des) [inline public/scripts/s3-uploads.js]: Not sure about these alerts ...

File: public/scripts/s3-uploads.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      }
+    }
+  };
+  xhr.send();
+}
+
+function uploadFile(file, signedRequest, url) {
+  const xhr = new XMLHttpRequest();
```
Comment (des-des) [inline public/scripts/s3-uploads.js:33]: With a helper

File: public/scripts/s3-uploads.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  xhr.send(file);
+}
+
+function saveInLocalStorage(imageId) {
+  localStorage.setItem('addedImage', imageId);
+}
+
+document.getElementById('photo-save').onclick = function (e) {
+  e.preventDefault();
+  if (localStorage.getItem('addedImage')) {
```
Comment (des-des) [inline public/scripts/s3-uploads.js]: Okay, this key represents if an upload is in progress.

Two things.
  1. I do not see why you need local storage for this.
  2. Is there a case where you are uploading more than 1 image / asset. Not really sure how you want to handle this
  3. You should call this imageUploadPending, and set it to true.

File: public/scripts/s3-uploads.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    const tag = $('.tag-input--photo')[0].value;
+    const heading = $('.heading-input--photo')[0].value;
+    const imageId = localStorage.getItem('addedImage');
+    $.ajax({
+      method: 'PUT',
+      url: 'memory-input-photo',
+      data: { tag, heading, imageId },
+      success: localStorage.removeItem('addedImage'),
```
Comment (des-des) [inline public/scripts/s3-uploads.js]: I expect you mean
```js
success: () => localStorage.removeItem('addedImage')
```

File: src/database/db_get.js [hunk 7]
```diff
@@ -43,8 +43,16 @@ const getUser = (input, callback) => {
   });
 };
 
+const getMemoryUrlById = (id, callback) => {
```
Comment (des-des) [inline src/database/db_get.js]: Why cant you just use getMemoryById?

File: src/database/db_update.js [hunk 8]
```diff
@@ -5,6 +5,15 @@ const updateLikes = (likeNum, memoryId, cb) => {
     'UPDATE memories SET likes = $1 WHERE memories.id = $2', [likeNum, memoryId], cb);
 };
 
+const updatePhotoMemory = (userId, tag, heading, imageId, cb) => {
+  connect.query(
+    `UPDATE memories
+     SET tag = $2, heading = $3
+     WHERE user_id = (SELECT users.id FROM users WHERE username = $1 OR email = $1)
```
Comment (des-des) [inline src/database/db_update.js]:
```sql
WHERE user_id = (SELECT users.id FROM users WHERE username = $1 OR email = $1)
```
is the same as
```sql
WHERE username = $1 OR email = $1
```
?

I am fairly confused, this set the heading and tag of a memory, using it image id and user id?

Comment (des-des) [reply src/database/db_update.js]: Hmm, why cant you just store the user Id in the session?

File: src/server/controllers/addPhotoMemory.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+const { updatePhotoMemory } = require('./../../database/db_update');
+
+module.exports = (req, res) => {
+  updatePhotoMemory('test', req.body.tag, req.body.heading, req.body.imageId, (err, res) => {
```
Comment (des-des) [inline src/server/controllers/addPhotoMemory.js]: Your endpoints are authenticated??

This is problematic

File: src/server/controllers/getPhotoMemory.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
+const aws = require('aws-sdk');
+require('env2')('./config.env');
+
+const S3_BUCKET_NAME = process.env.S3_BUCKET_NAME;
+
+module.exports = (req, res) => {
+  aws.config.update({ accessKeyId: process.env.AWSAccessKeyId, secretAccessKey: process.env.AWSSecretKey, signatureVersion: 'v4', region: 'eu-west-2' });
```
Comment (des-des) [inline src/server/controllers/getPhotoMemory.js]: Bring this out into its own helper function

File: src/server/controllers/getPhotoMemory.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  const s3 = new aws.S3();
+
+  const s3Params = {
+    Bucket: S3_BUCKET_NAME,
+    Key: req.query.memoryUrl,
```
Comment (des-des) [inline src/server/controllers/getPhotoMemory.js]: Okay, there are 2 things you are missing here.

Firstly, you need to check if this memory belongs to the user!

I would

  1. Only keep the memoryId on the client (not the s3 key, helpfully called memoryUrl)
  2. In the endpoint, use the id from the session (this means that the user has been authenticated), and the memoryId, to find a the s3key for the memory.
  3. Make sure you check the memory belongs to the authenticated user!

File: src/server/controllers/getPhotoMemory.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  };
+
+  s3.getSignedUrl('getObject', s3Params, (err, url) => {
+    if (err) {
+      console.log(err);
+      return res.end();
```
Comment (des-des) [inline src/server/controllers/getPhotoMemory.js]: this is unhelpful. what does the client do??
```js
res.status(500).send('failed to generate upload url')
```

File: src/server/controllers/signS3.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+const connect = require('./../../database/db_connect.js');
+
+const cuid = require('cuid');
+const { createMemory } = require('./../../database/db_create');
+
+const S3_BUCKET_NAME = process.env.S3_BUCKET_NAME;
+
+module.exports = (req, res) => {
+  aws.config.update({ accessKeyId: process.env.AWSAccessKeyId, secretAccessKey: process.env.AWSSecretKey, signatureVersion: 'v4', region: 'eu-west-2' });
```
Comment (des-des) [inline src/server/controllers/signS3.js]: Again, put the s3 init logic in a helper

File: src/server/controllers/getPhotoMemory.js (hunk 11 above)
Comment (des-des) [inline src/server/controllers/getPhotoMemory.js]: Okay,

  1. This is the s3 key, you should not give this to the client.
  2. The client should send the memoryId
  3. Use the memoryId to find the s3 key
  4. **Check that this memory belongs to the user**. Ie query the memories against the memoryId **and** the userId, to retrieve the s3Key

File: src/server/controllers/getPhotoMemory.js (hunk 12 above)
Comment (des-des) [inline src/server/controllers/getPhotoMemory.js]: res.status(500).send('failed to generate asset URL')

File: src/server/controllers/signS3.js [hunk 14]
```diff
@@ -0,0 +1,49 @@
+const express = require('express');
```
Comment (des-des) [inline src/server/controllers/signS3.js]: Why is this called signS3, this is confusing.

I think you should call the files

`assetCreateUrl`, `assetGetUrl`, `assetDeleteUrl`.

Right now the naming is all over the place

File: src/server/controllers/signS3.js (hunk 13 above)
Comment (des-des) [inline src/server/controllers/signS3.js]: Put this in a helper..

File: src/server/controllers/signS3.js [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    ACL: 'public-read',
+  };
+
+  const newMemory = {
+    memory_text: '',
+    heading: '',
+    tag: '',
+  };
+
+  createMemory('test', newMemory, 'image', fileKey, (err, res) => {
```
Comment (des-des) [inline src/server/controllers/signS3.js]: Hmm, this is not how you do async in js..
```js
(res, res) => {
  db.createMemory(res.body, err => {
    if (err) return res.status(500).send(message bla)
    s3.getSignedUrl(() => {
      res.success ..
    })
  })
}
```js

File: src/server/controllers/signS3.js [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    if (err) return err;
+  });
+
+  s3.getSignedUrl('putObject', s3Params, (err, data) => {
+    if (err) {
+      console.log(err);
+      return res.end();
```
Comment (des-des) [inline src/server/controllers/signS3.js]: Why no status code, no message ..

File: src/server/controllers/signS3.js [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    }
+    const returnData = {
+      key: s3Params.Key,
+      signedRequest: data,
+      url: `https://${S3_BUCKET_NAME}.s3.amazonaws.com/${fileName}`,
+    };
+    res.write(JSON.stringify(returnData));
```
Comment (des-des) [inline src/server/controllers/signS3.js]:
```js
res.json(returnData)
```

**Review by des-des — CHANGES_REQUESTED** [review]
okay, need some work ...

But yeah congrats getting this working, I know its a real pain

File: public/scripts/s3-uploads.js (hunk 6 above)
Comment (joeylouise) [reply public/scripts/s3-uploads.js]: I've changed this to

`success: () => { imageUploadPending = false; }`

File: src/database/db_get.js (hunk 7 above)
Comment (joeylouise) [reply src/database/db_get.js]: Have renamed this to getMemoryById

File: src/server/controllers/getPhotoMemory.js (hunk 10 above)
Comment (joeylouise) [reply src/server/controllers/getPhotoMemory.js]: I've created an `s3init` function

File: src/server/controllers/signS3.js (hunk 13 above)
Comment (joeylouise) [reply src/server/controllers/signS3.js]: I've created an `s3init` function

Comment (joeylouise) [reply src/server/controllers/signS3.js]: I've created an `s3init` function

File: src/server/controllers/signS3.js (hunk 15 above)
Comment (joeylouise) [reply src/server/controllers/signS3.js]: I've refactored this 👍

File: src/server/controllers/signS3.js (hunk 17 above)
Comment (joeylouise) [reply src/server/controllers/signS3.js]: I've changed this

File: src/server/controllers/getPhotoMemory.js (hunk 12 above)
Comment (joeylouise) [reply src/server/controllers/getPhotoMemory.js]: I've added this in

Comment (joeylouise) [reply src/server/controllers/getPhotoMemory.js]: Added `res.status(500).send('failed to generate asset URL')`

File: src/database/db_update.js (hunk 8 above)
Comment (joeylouise) [reply src/database/db_update.js]: Cool, now storing user id in session as well
#152

Comment (joeylouise) [reply src/database/db_update.js]: Passing in user.id now so have changed query to 

`WHERE user_id = $1`

File: public/scripts/s3-uploads.js (hunk 5 above)
Comment (joeylouise) [reply public/scripts/s3-uploads.js]: I've removed where we use localStorage and instead are using `imageUploadPending` variable

File: src/server/controllers/signS3.js (hunk 16 above)
Comment (joeylouise) [reply src/server/controllers/signS3.js]: Added status code and message

File: src/server/controllers/addPhotoMemory.js (hunk 9 above)
Comment (joeylouise) [reply src/server/controllers/addPhotoMemory.js]: Have changed test to `req.session.id`
Need to add authentication to all routes

**Comment by des-des** [comment]
@joeylouise push?

**Comment by joeylouise** [comment]
@des-des ok...have pushed some stuff, still working on it

**Comment by des-des** [comment]
ah cool

File: src/server/controllers/getPhotoMemory.js (hunk 11 above)
Comment (joeylouise) [reply src/server/controllers/getPhotoMemory.js]: k have refactored this now so that memory_asset_url is not kept on client

Comment (joeylouise) [reply src/server/controllers/getPhotoMemory.js]: have done this now

**Comment by joeylouise** [comment]
Hey @des-des I've made a bunch of changes to this, there are a couple of things I'm not sure about (making the xhr function in a helper and what to change the `alert`s to. Shall I assign to you to review so far or should we have a chat first?

**Comment by des-des** [comment]
@joeylouise since we are both in the space lets just chat. Come grab me whenever

**Comment by joeylouise** [comment]
Hey @des-des, I tried to refactor the xhr requests and everything kept breaking and I didn't want to take any more time refactoring so I have refactored a little bit and got rid of all alerts...

File: src/database/db_get.js [hunk 18]
```diff
@@ -43,8 +43,16 @@ const getUser = (input, callback) => {
   });
 };
 
+const getMemoryById = (userId, id, callback) => {
+  connect.query('SELECT memory_asset_url FROM memories WHERE user_id = $1 AND id = $2', [userId, id], (err, res) => {
+    if (err) return err;
```
Comment (des-des) [inline src/database/db_get.js]: :scream: the callback is not being called when you have an error

File: src/server/app.js [hunk 19]
```diff
@@ -5,6 +5,9 @@ const path = require('path');
 const bodyParser = require('body-parser');
 const cookieSession = require('cookie-session');
 const router = require('./router');
+const aws = require('aws-sdk');
```
Comment (des-des) [inline src/server/app.js]: does this do anything?

File: src/server/app.js [hunk 20]
```diff
@@ -5,6 +5,9 @@ const path = require('path');
...
+
+const S3_BUCKET_NAME = process.env.S3_BUCKET;
```
Comment (des-des) [inline src/server/app.js]: and this?

File: src/server/controllers/assetCreateUrl.js [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+        return res.status(500).send('failed to generate asset URL');
+      }
+      const returnData = {
+        imageId: response.rows[0].id,
+        key: s3Params.Key,
+        signedRequest: data,
+        url: `https://${S3_BUCKET_NAME}.s3.amazonaws.com/${fileName}`,
+      };
+      res.json(returnData);
+      res.end();
```
Comment (des-des) [inline src/server/controllers/assetCreateUrl.js]: Do you need to call res.json and res.end?

File: src/server/controllers/assetDeleteUrl.js [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+      },
+    };
+
+    s3.deleteObjects(params, (err, data) => {
+      if (err) console.log(err);
+      else console.log(data);
+    });
+  });
+
+  deleteMemory(req.body.id, (error, response) => {
```
Comment (des-des) [inline src/server/controllers/assetDeleteUrl.js]: You should nest all you async stuff, right now you are running s3.delete object and deleteMemory in parallel.

Put one inside the other, or use async.parallel.

File: src/server/controllers/assetDeleteUrl.js [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+          {
+            Key: response,
+          },
+        ],
+      },
+    };
+
+    s3.deleteObjects(params, (err, data) => {
+      if (err) console.log(err);
+      else console.log(data);
```
Comment (des-des) [inline src/server/controllers/assetDeleteUrl.js]: not sure you want to log this

**Review by des-des — CHANGES_REQUESTED** [review]

**Comment by joeylouise** [comment]
@des-des cool. 👍

**Review by des-des — APPROVED** [review]

**Pull request merged**